### ISSUE TITLE
Data widgets errors

### ISSUE THREAD

**Issue by jhpedemonte** [maqetta/maqetta#1764]
I've noticed two serious bugs with "data" widgets (DataGrid, Tree, RoundRectDataList, EdgeToEdgeDataList).
1. Nothing happens when first added to page.  There is no error, not even if I set debugger to break on all errors.  Just silently fails to add the widget to the page. Adding widget a 2nd time works.
2. Setting new values through smart input leads to error.  Seems the widget code cannot find the data store.

@peller

**Comment by peller** [comment]
Sounds like it's somehow related to the async changes.

**Comment by billreed63** [comment]
The widget helpers are not being loaded. they need to be loaded before we attempt to Modify the widgets.

**Comment by billreed63** [comment]
4:20:18 PM: Adam L. Peller: just looking at the store data bug
4:20:27 PM: Adam L. Peller: looks like an async getHelper problem
4:20:32 PM: William D. Reed: It is the helpers are not getting loaded
4:20:36 PM: Adam L. Peller: yah
4:20:52 PM: Adam L. Peller: for the bootstrapping case
4:20:55 PM: William D. Reed: _preProcess: function (node){
        //need a helper to pre process widget
        // also, prime the helper cache
        var type = node.getAttribute("dojoType");
        return Widget.requireWidgetHelper(type).then(function(helper) {  
            if(helper && helper.preProcess){
                helper.preProcess(node, this);
            }
        });
    },
4:20:58 PM: Adam L. Peller: though it's odd, because there's already a tree here
4:21:02 PM: William D. Reed: this code never gets called
4:21:08 PM: Adam L. Peller: hmm
4:21:23 PM: William D. Reed: It only gets called if the page has widgets on it...
4:21:32 PM: Adam L. Peller: yeah
4:21:38 PM: Adam L. Peller: not for the widgets you add
4:21:45 PM: William D. Reed: And it does not get called if the widget is a datagrid
4:21:52 PM: Adam L. Peller: in this case, it's not even the widget, it's the store
4:22:04 PM: Adam L. Peller: how to do that asynchronously...
4:22:44 PM: Adam L. Peller: what if the tool pulled in its dependencies?
4:22:45 PM: William D. Reed: Maybe when the widget palette is populated
4:22:52 PM: Adam L. Peller: nah
4:23:04 PM: Adam L. Peller: would be expensive
4:23:10 PM: William D. Reed: the create tool, that would work
4:23:32 PM: William D. Reed: it has to happen before the create...
4:24:05 PM: Adam L. Peller: if TreeCreateTool had a dependency on the necessary module, perhaps that would work
4:24:21 PM: Adam L. Peller: there was another bug like this? the combobox one?
4:24:28 PM: Adam L. Peller: and datagrid?
4:24:32 PM: William D. Reed: no, some widgets don't have create tools but may have helpers
4:24:40 PM: Adam L. Peller: hrm
4:24:44 PM: William D. Reed: it is all the same bug...
4:25:13 PM: William D. Reed: when the widget get DnD from the palette the helper needs to be loaded
4:25:33 PM: William D. Reed: maybe in ve/CreatTool
4:25:38 PM: Adam L. Peller: must be similar to the PasteTool path
4:25:41 PM: Adam L. Peller: but not the same code
4:26:50 PM: Adam L. Peller: so when I asynchronously request the tool I can asynchronously request the helper at the same time
4:26:53 PM: Adam L. Peller: that shouldn't be too hard
4:27:03 PM: Adam L. Peller: problem is, some like tree require other helpers
4:27:17 PM: Adam L. Peller: I guess that logic would be specific to the createtool?
4:27:36 PM: William D. Reed: True but they could add them in their CreatTools
4:27:43 PM: Adam L. Peller: but where
4:27:55 PM: Adam L. Peller: we'd need to make another Deferred method to call
4:28:00 PM: Adam L. Peller: on the tool
4:28:31 PM: Adam L. Peller: does that make sense?
4:29:32 PM: William D. Reed: We could fire and forget.. it is just priming the Helper cache
4:30:01 PM: William D. Reed: maybe...
4:30:31 PM: Adam L. Peller: unfortunately, no… if we don't wait, we'll be in a race condition here
4:30:53 PM: Adam L. Peller: but I suspect there are lots of places like that in the code now where we do unintentionally fire and forget, and if the user is too quick to do something… bam
4:31:20 PM: Adam L. Peller: I can chain all of these methods together in PasteTool._create and the other CreateTool
4:31:27 PM: Adam L. Peller: hopefully those are the only two?
4:31:37 PM: Adam L. Peller: ways to create a widget
4:31:46 PM: William D. Reed: delete also
4:31:46 PM: Adam L. Peller: the third being document loading (_preProcess)
4:31:48 PM: William D. Reed:  
4:31:51 PM: Adam L. Peller: well
4:31:56 PM: Adam L. Peller: by then, it should have been loaded?
4:32:43 PM: William D. Reed: True

**Comment by peller** [comment]
see #1768

**Comment by billreed63** [comment]
Working now
